debug print convenience functions

kd0aij opened the pull request:
allows specifiying print interval and max number of prints

uses no resources unless a function is invoked

kd0aij commented:
@khancyr Thanks!
Is there a way to #include DBGprint.h so that it's available everywhere?

khancyr commented:
I would say in AP_Logger or GCS because of the dependency on those and they are heavily used !

kd0aij commented:
Moved files to GCS_MAVLink and added include in GCS.h
Example demos all 3 dbgprint options (PRINTF, CONSOLE, GCS).
dbgEdgePrint and dflog have been tested in a Plane build.

kd0aij commented:
@khancyr Thanks for the advice. The CI errors seem to be unrelated to this PR, and the functions are working properly for me locally.

kd0aij commented:
@khancyr Managed to fix the compilation errors by rearranging the #include statements.

peterbarker commented on libraries/GCS_MAVLink/DBGprint.h:
```diff
@@ -0,0 +1,74 @@
+
+class DBGprint {
+public:
+    enum class Otype {
```
Probably should be `Otype`

kd0aij replied:
did you mean OType?

kd0aij commented:
@peterbarker changed Otype to OType and squashed
I hope you find this useful :)

peterbarker commented:
On Mon, 8 Jun 2020, Mark Whitehorn wrote:

> +class DBGprint {
> +public:
> +    enum class Otype {
> 
> did you mean OType?

Yes.

khancyr commented:
@kd0aij that is normal, as you don't call the function anywhere.

kd0aij commented:
I was also wondering whether there might be a compilation time penalty due to including DBGprint.h in GCS.h.

kd0aij commented:
ran astyle and force-pushed

kd0aij commented:
added millis() timestamp at front of printout. 
**Output from ./build/sitl/examples/DBGprint -C --model plane:**
```
Starting sketch 'UNKNOWN'
Starting SITL input
DBGprint tests

construct new DBGprint instance 0: 7ff392c0e38f
1000: printf 1000
construct new DBGprint instance 1: 7ff392c0e3b7
construct new DBGprint instance 2: 7ff392c0e3df
construct new DBGprint instance 3: 7ff392c0e455
1000: watch, value: 1, time: 1000
1000: console 1000
TOGCS: gcs 1000

1250: printf 1250
1500: printf 1500
1500: console 1500
TOGCS: gcs 1500

1750: printf 1750
2000: printf 2000
2000: watch, value: 0, time: 2000
2000: console 2000
TOGCS: gcs 2000
```
**Debug output from pr-tilt-disarmed:**
mavproxy console on arming (note that reset_delay_arming occurs ~2 seconds after arming)
```
arm throttle
QSTABILIZE> APM: Throttle armed
Got COMMAND_ACK: COMPONENT_ARM_DISARM: ACCEPTED
ARMED
Arming checks disabled
29324: reset_delay_arming
```
gdb console on arming:
```
construct new DBGprint instance 0: 80649be
construct new DBGprint instance 1: 8057057
27331: delaying spoolup
28331: delaying spoolup
construct new DBGprint instance 2: 80570b0
```
mavproxy console on disarming (note that tilt is delayed for 3 seconds after disarm):
```
disarm
QSTABILIZE> APM: Throttle disarmed
Got COMMAND_ACK: COMPONENT_ARM_DISARM: ACCEPTED
96404: delaying tilt
DISARMED
97404: delaying tilt
98410: delaying tilt
```

kd0aij commented:
SITL retest with name of caller added:
```
QSTABILIZE> arm throttle
22291: arm: set delay_arming
APM: Throttle armed
Got COMMAND_ACK: COMPONENT_ARM_DISARM: ACCEPTED
22298: motors_output: delaying spoolup
ARMED
Arming checks disabled
23298: motors_output: delaying spoolup
24291: update_soft_armed: reset delay_arming
APM: EKF yaw reset -2.79
QSTABILIZE> disarm
APM: Throttle disarmed
Got COMMAND_ACK: COMPONENT_ARM_DISARM: ACCEPTED
37796: tiltrotor_vectored_yaw: delaying tilt
DISARMED
38796: tiltrotor_vectored_yaw: delaying tilt
39796: tiltrotor_vectored_yaw: delaying tilt
```
